domain/secret/watcher_test.go

```diff
@@ -189,42 +170,66 @@ func (s *watcherSuite) TestWatchObsoleteForAppsAndUnitsOwned(c *gc.C) {
 		},
 	)
 	c.Assert(err, gc.IsNil)
-	c.Assert(watchAll, gc.NotNil)
-	defer workertest.CleanKill(c, watchAll)
-
-	wCAll := watchertest.NewStringsWatcherC(c, watchAll)
-
-	// Wait for the initial changes.
-	wCAll.AssertChange([]string(nil)...)
-	s.AssertChangeStreamIdle(c)
-
-	// create revision 2, and obsolete revision 1.
-	createNewRevision(c, st, uri1)
-	createNewRevision(c, st, uri2)
-	createNewRevision(c, st, uri3)
-	createNewRevision(c, st, uri4)
-
-	s.AssertChangeStreamIdle(c)
-	wCAll.AssertChange(
-		revID(uri1, 1),
-		revID(uri2, 1),
-		revID(uri3, 1),
-		revID(uri4, 1),
-	)
+	c.Assert(w, gc.NotNil)
+	defer watchertest.CleanKill(c, w)
+
+	harness := watchertest.NewHarness(s, watchertest.NewWatcherC(c, w))
+	harness.AddTest(func(c *gc.C) {
+		sp := secret.UpsertSecretParams{
+			Data: coresecrets.SecretData{"foo": "bar", "hello": "world"},
+		}
+		sp.RevisionID = ptr(uuid.MustNewUUID().String())
+		err := createCharmApplicationSecret(ctx, st, 1, uri1, "mysql", sp)
+		c.Assert(err, jc.ErrorIsNil)
+
+		sp.RevisionID = ptr(uuid.MustNewUUID().String())
+		err = createCharmUnitSecret(ctx, st, 1, uri2, "mysql/0", sp)
+		c.Assert(err, jc.ErrorIsNil)
 
-	// create revision 3, and obsolete revision 2.
-	createNewRevision(c, st, uri1)
-	createNewRevision(c, st, uri2)
-	createNewRevision(c, st, uri3)
+		sp.RevisionID = ptr(uuid.MustNewUUID().String())
+		err = createCharmApplicationSecret(ctx, st, 1, uri3, "mediawiki", sp)
+		c.Assert(err, jc.ErrorIsNil)
 
-	s.AssertChangeStreamIdle(c)
-	wCAll.AssertChange(
-		revID(uri1, 2),
-		revID(uri2, 2),
-		revID(uri3, 2),
-	)
+		sp.RevisionID = ptr(uuid.MustNewUUID().String())
+		err = createCharmUnitSecret(ctx, st, 1, uri4, "mediawiki/0", sp)
+		c.Assert(err, jc.ErrorIsNil)
+	}, func(w watchertest.WatcherC[[]string]) {
+		w.AssertNoChange()
+	})
+
+	harness.AddTest(func(c *gc.C) {
+		// create revision 2, and obsolete revision 1.
```

Comment:
The comments here should be placed on top of AddTest and describe better the affects on the system and the side effects we expect to see. i.e document the business logic we are expecting to see here. Otherwise these tests don'e mean a great deal on their own. It's not easy to grok what we are trying to test.